arcgis-android-toolkit/src/main/java/com/esri/arcgisruntime/toolkit/ar/ArcGISArView.kt

```diff
@@ -163,18 +243,131 @@ final class ArcGISArView : FrameLayout, DefaultLifecycleObserver, Scene.OnUpdate
+                if (initialLocation == null) {
+                    initialLocation = location
+                    cameraController.originCamera = Camera(location, 0.0, 0.0, 0.0)
+                } else if (isUsingARCore != ARCoreUsage.YES) {
+                    val camera = sceneView.currentViewpointCamera.moveTo(location)
+                    sceneView.setViewpointCamera(camera)
+                }
+            }
+        }
+
+    /**
+     * This listener is added to every [LocationDataSource] used when using the [locationDataSource] property to receive
+     * heading updates.
+     *
+     * Upon receiving an updated heading, if we're not using ARCore we create a [Camera] using the updated heading value
+     * and the pitch and roll of the current Camera from [sceneView]. We then set the current Camera of the [SceneView]
+     * to be the newly created Camera.
+     *
+     * @since 100.6.0
+     */
+    private val headingChangedListener: LocationDataSource.HeadingChangedListener =
+        LocationDataSource.HeadingChangedListener {
+            if (isUsingARCore != ARCoreUsage.YES && !it.heading.isNaN()) {
+                // Not using ARCore, so update heading on the camera directly; otherwise, let ARCore handle heading changes.
+                val currentCamera = sceneView.currentViewpointCamera
+                val camera = currentCamera.rotateTo(it.heading, currentCamera.pitch, currentCamera.roll)
+                sceneView.setViewpointCamera(camera)
+            }
+        }
+
+    /**
+     * This listener is added to every [LocationDataSource] used when using the [locationDataSource] property to receive
+     * LocationDataSource status updates.
+     *
+     * Upon receiving the status of the LocationDataSource, we check if the status indicates a failure. If so, we set the
+     * [error] property to an Exception that indicates the current error of the LocationDataSource.
+     *
+     * @since 100.6.0
+     */
+    private val locationDataSourceStatusChangedListener: LocationDataSource.StatusChangedListener =
+        LocationDataSource.StatusChangedListener {
+            if (it.status == LocationDataSource.Status.FAILED_TO_START) {
+                error = Exception(locationDataSource?.error)
+                isTracking = isUsingARCore == ARCoreUsage.YES
+            }
+        }
+
+    /**
+     * The data source used to get device location.  Used either in conjunction with ARCore data or when ARCore is not
+     * present or not being used.
+     *
+     * @since 100.6.0
+     */
+    var locationDataSource: LocationDataSource? = null
+        set(value) {
+            isTracking = if (value == null) {
+                field?.removeLocationChangedListener(locationChangedListener)
+                field?.removeHeadingChangedListener(headingChangedListener)
+                field?.removeStatusChangedListener(locationDataSourceStatusChangedListener)
+                isUsingARCore == ARCoreUsage.YES
+            } else {
```

Comment:
if `value` is not null, but we have a `locationDataSource` set previously, we still should unsubscribe the listeners. In other words, if `field` is not null, we should always unsubscribe the listeners

Reply:
@LDuncAndroid - this comment still needs to be addressed

Reply:
Will fix.